src/witness/yamlWitness.ml

```diff
@@ -174,43 +218,107 @@ struct
             | None ->
               acc
           end
-        | _ -> (* avoid FunctionEntry/Function because their locations are not inside the function where assert could be inserted *)
+        end else begin
+          (* avoid FunctionEntry/Function because their locations are not inside the function where assert could be inserted *)
           acc
+        end
       ) nh []
     in
 
-    (* TODO: deduplicate *)
+
+    (* Generate precondition invariants.
+       We do this in three steps:
+       1. Collect contexts for each function
+       2. For each function context, find "matching"/"weaker" contexts that may satisfy its invariant
+       3. Generate precondition invariants. The postcondition is a disjunction over the invariants for matching states. *)
+
+    (* 1. Collect contexts for each function *)
+    let fun_contexts : con_inv list FMap.t = FMap.create 103 in
+    LHT.iter (fun (n, c) local ->
+        begin match n with
+          | FunctionEntry f ->
+            let context = node_context n in
+            let invariant = Spec.D.invariant context local in
+            FMap.modify_def [] f (fun acc -> {context = c; invariant; node = n; state = local}::acc) fun_contexts
+          | _ -> ()
+        end
+      ) lh;
+
+    (* 2. For all contexts and their invariants, find all contexts such that their start state may satisfy the invariant. *)
+    let fc_map : con_inv list FCMap.t = FCMap.create 103 in
+    FMap.iter (fun f con_invs ->
+        List.iter (fun current_c ->
+            (* Collect all start states that may satisfy the invariant of current_c *)
+            List.iter (fun c ->
+                begin match current_c.invariant with
+                  | Some c_inv ->
+                    let x = ask_local (c.node, c.context) c.state (Queries.EvalInt c_inv) in
+                    if Queries.ID.is_bot x || Queries.ID.is_bot_ikind x then (* dead code *)
+                      failwith "Bottom not expected when querying context state" (* Maybe this is reachable, failwith for now so we see when this happens *)
+                    else if Queries.ID.to_bool x = Some false then () (* Nothing to do, the c does definitely not satisfy the predicate of current_c *)
+                    else begin
+                      (* Insert c into the list of weaker contexts of f *)
+                      FCMap.modify_def [] (f, current_c.context) (fun cs -> c::cs) fc_map;
+                    end
+                  | None ->
+                    (* If the context invariant is None, we will not generate a precondition invariant. Nothing to do here. *)
```

Comment:
Is there a reason to even add them into this quadratic checking loop then?

Reply:
Good point, now I first check whether `current_c.invariant` is an actual expression, before iterating over all the other contexts of the function.
Such contexts cannot be excluded from the inner iteration though; thus I still keep them in the `fun_contexts` data structure.